[feature] Made Device verbose_name configurable #287

codesankalp opened the pull request:
Added configurable verbose name field for Device name in the meta
class of AbstractDevice in device.py.

This value for this verbose_name is being imported from variable
OPENWISP_DEVICE_VERBOSE_NAME defined in settings.py.

Fixes #287

codesankalp commented:
I have tried many times the checks are passing in my device but fails in travis.
I think this is due to unable to apply migrations while running sample app.

NoumbissiValere commented:
@codesankalp the test is still failing because you've not ran `makemigrations` on `SAMPLE_APP`. This can be achieved using the commands `SAMPLE_APP=1 ./manage.py makemigrations`.

NoumbissiValere commented:
But I doubt if this is the best way to go because each time the settings is updated with a new verbose_name, one will need to `makemigrations` again. I am thinking it will be best to update this at the `admin`

NoumbissiValere commented on openwisp_controller/config/settings.py:
```diff
@@ -52,3 +52,4 @@ def get_settings_value(option, default):
 }
 HARDWARE_ID_OPTIONS.update(get_settings_value('HARDWARE_ID_OPTIONS', {}))
 HARDWARE_ID_AS_NAME = get_settings_value('HARDWARE_ID_AS_NAME', True)
+OPENWISP_DEVICE_VERBOSE_NAME = 'device'
```
Please follow the way other settings have been defined. This could be written as `DEVICE_VERBOSE_NAME = get_settings_value('DEVICE_VERBOSE_NAME',  'Device')

codesankalp replied:
ok, @NoumbissiValere  I will update the readme as well as do the mentioned changes.
Thanks.

codesankalp replied:
Also, i want to know how to pass these QA-checks because while running locally it is passing but in Travis, it is failing.

NoumbissiValere replied:
You can do that with the command `./run-qa-checks`

NoumbissiValere reviewed (changes requested):
We will also need to update the README as well wherw other settings have been specified and explain what this setting is use for.

codesankalp commented:
@NoumbissiValere please review it again.
Thanks.

NoumbissiValere commented on README.rst:
```diff
@@ -563,6 +563,22 @@ their hardware ID instead of their name.
+``Network Configuration administration``.
+
+You can use more specialized word rather than generic device.
+
```
Please remove this extra newlines

codesankalp replied:
I followed the format of the readme that's why i have added new lines.

NoumbissiValere replied:
You are right @codesankalp but we need to know where to use this newlines. Generally, newlines are used to separate paragraphs or new ideas and not sentences. Let's keep the documentation simple. we could write something like:

```
Defines the verbose_name attribute of Device model, which is displayed on admin site.
For example, if we want to change the verbose name of Device model at admin, we could write;

.. code-block: python

    OPENWISP_CONTROLLER_DEVICE_VERBOSE_NAME = 'Router'
```

NoumbissiValere commented on openwisp_controller/config/base/device.py:
```diff
@@ -83,6 +83,7 @@ class Meta:
         )
         abstract = True
+        verbose_name = app_settings.DEVICE_VERBOSE_NAME
 
```
I don't think this change should be done in a model. Because each time you set a new value to the settings, you will need to run `./manage.py makemigrations` and `./manage.py migrate` and this won't be good at production. I suggest we look into other better solution like doing something in the `admin.py` or @nemesisdesign what do you say?

pandafy replied:
@NoumbissiValere I think https://github.com/openwisp/openwisp-controller/pull/296#discussion_r499215099 will solve this issue. What are your thought?

pandafy commented on openwisp_controller/config/migrations/0033_configurable_verbose_device_name.py:
```diff
@@ -0,0 +1,16 @@
+
+    operations = [
+        migrations.AlterModelOptions(
+            name='device', options={'verbose_name': 'Device'},
```
Here we can import `app_settings` and change this line to 
```
name='device', options={'verbose_name': app_settings.DEVICE_VERBOSE_NAME},
```

This will not create any new migrations when `OPENWISP_CONTROLLER_DEVICE_VERBOSE_NAME` is configured by user.

codesankalp replied:
I solved this problem using different approach I will add that commit soon.
In my method there is no need of migration file.

pandafy commented on README.rst:
```diff
@@ -563,6 +563,22 @@ their hardware ID instead of their name.
+| **default**: | ``Device``  |
++--------------+-------------+
+
+When enabled, makes device verbose name configurable in the
+``Network Configuration administration``.
```
@codesankalp this is setting is not something which can be enabled or disabled like a boolean flag. This configures another setting.  This should be something along the lines of 
```
Defines the verbose_name attribute of Device model, which is displayed on admin site. 
```

codesankalp commented:
@pandafy , @NoumbissiValere I think after this commit there is no need of make migrations.

NoumbissiValere commented on README.rst:
```diff
@@ -563,6 +563,21 @@ their hardware ID instead of their name.
 
 If you still want to reference devices by their name, set this to ``False``.
 
+``OPENWISP_CONTROLLER_DEVICE_VERBOSE_NAME``
```
Please update your commit message. your are talking of `OPENWISP_DEVICE_VERBOSE_NAME` in your commit message. it should be `OPENWISP_CONTROLLER_DEVICE_VERBOSE_NAME` as here

NoumbissiValere commented on openwisp_controller/config/admin.py:
```diff
@@ -52,6 +52,11 @@
 else:  # pragma: nocover
     from django.contrib.admin import ModelAdmin
 
+Device._meta.verbose_name = app_settings.get_settings_value(
```
This will work and i think it is a better solution. But please define the setting where it was and use it as before
That is, define the `DEVICE_VERBOSE_NAME` in the setting file as you did it before and then here do
`Device._meta.verbose_name = app_settings.DEVICE_VERBOSE_NAME`

codesankalp replied:
``Device._meta.verbose_name = app_settings.DEVICE_VERBOSE_NAME``
is showing attribution error because the DEVICE_VERBOSE_NAME is not defined.

Should i use ``get_settings_value``
@NoumbissiValere 

Also i have to assign verbose_plural_name because it will not be automatically updated.

NoumbissiValere replied:
@codesankalp please you have to defined `DEVICE_VRRBOSE_NAME` in config/settings file like you did it before [here](https://github.com/openwisp/openwisp-controller/pull/296#discussion_r499156526)

NoumbissiValere replied:
According to your last commit, the `verbose_name_plural` depends on the `verbose_name` so any changes in your verbose name will be appropriately reflected in the plural

codesankalp replied:
@NoumbissiValere as I am changing the meta class externally it will not affect the plural one I have to manually add that too.

codesankalp replied:
I have checked your suggestion also but it didn't automatically update plurals.

NoumbissiValere replied:
@codesankalp  this is the code you wrote;
```
Device._meta.verbose_name = app_settings.get_settings_value(
     'DEVICE_VERBOSE_NAME', 'Device'
)
Device._meta.verbose_name_plural = Device._meta.verbose_name + 's'
```
Right? If i'm wrong, please tell me but that's the code i see in this last commit.
The code works but it is against how we build things on `openwisp`.

I am saying you should define the settings in the file where other settings are defined. you did this before. [here](https://github.com/openwisp/openwisp-controller/pull/296#discussion_r499156526) Then come here and simple do;
`Device._meta.verbose_name = app_settings.DEVICE_VERBOSE_NAME`. 
This will mean that your total new changes in the admin file will look like this;

```
Device._meta.verbose_name = app_settings.DEVICE_VERBOSE_NAME
Device._meta.verbose_name_plural = Device._meta.verbose_name + 's'
```
Please if it is not still clear, let me know

codesankalp replied:
Thanks, @NoumbissiValere I did the same and I am testing it currently.
Previously I thought you want me to remove `verbose_name_plural` and use only `verbose_name`.
That's my confusion thanks for clearing it.

NoumbissiValere reviewed (changes requested):
Great progress @codesankalp just some small touches and i think it will be good :smile:

codesankalp commented:
@NoumbissiValere can you please review it?

NoumbissiValere reviewed (changes requested):
Great work @codesankalp the code looks great now and things are in place. But there is no one who will always check that this feature is always like this. someone can mistakenly change it in 2 or 5 or even 10 years after and we would have even forgotten why we even did this in the first place. to avoid things like this, we use what we call an automated test. which will always check that this feature exist.  So the last thing we have to add to this PR to make it superb is a test case.

you can add a test case in `config/tests/test_admin.py` that checks that when the settings is updated with another value, this is reflected on the admin site. you can do this by checking that the new verbose name can be seen in `device changelist`.

If you have any doubt, please ask

codesankalp commented:
@NoumbissiValere I am doing the above-mentioned task but while changing the device name from setting it is not changing the verbose. Can you help me?

codesankalp commented:
I am facing problem because when i change it in test_admin using settings it do not reflects in the actual response.

NoumbissiValere commented:
@codesankalp please can you send what you have done so far? So that I can know how to direct and help you?

codesankalp commented:
![1](https://user-images.githubusercontent.com/56037184/95097076-b0326880-074a-11eb-9674-1b446e93c8c6.png)
![2](https://user-images.githubusercontent.com/56037184/95097085-b1fc2c00-074a-11eb-90cd-2fc06528c2c6.png)
![3](https://user-images.githubusercontent.com/56037184/95097087-b1fc2c00-074a-11eb-8d01-f193df8bd35d.png)
![4](https://user-images.githubusercontent.com/56037184/95097091-b294c280-074a-11eb-833c-a77133635bf3.png)
these is the present code, in morning I used the `device_changelist` also but I removed that because it is not updating in device so i removed that line for get response and assertion for now.

codesankalp commented:
@NoumbissiValere please help me in this.

pandafy commented:
@codesankalp **be patient with reviews.** 

Read the documentation for [modify_settings in Django's documentation](https://docs.djangoproject.com/en/3.1/topics/testing/tools/#django.test.SimpleTestCase.modify_settings). Does it suits your use case? 

Check out what alternatives are present in the documentation i have shared. Look at the sections before and after the linked documentation. You will find your answer, if you don't let me know I will share it with you. Always read about things you are using before jumping onto a conclusion. 

https://github.com/openwisp/openwisp-notifications/blob/1d9c8de2b7c250f4b600c15dd3c6f89fa998d423/openwisp_notifications/tests/test_admin.py#L248-L261

nemesisdesign commented on openwisp_controller/config/admin.py:
```diff
@@ -52,6 +52,9 @@
     from django.contrib.admin import ModelAdmin
 
+Device._meta.verbose_name = app_settings.DEVICE_VERBOSE_NAME
+Device._meta.verbose_name_plural = Device._meta.verbose_name + 's'
```
this should go here: https://github.com/openwisp/openwisp-controller/blob/master/openwisp_controller/config/base/device.py#L79-L85

I see that you had an interaction with @NoumbissiValere and @pandafy about it, but this makes the code harder for a developer working with OpenWISP, they will have a hard time finding these two lines and then they'll be pretty surprised to find these here.
The right approach is to define this at model level and edit the migration so the migration also uses the setting instead of the raw value.

It would also be better to use a tuple in which the first element is the singular form and the second one the plural form, just to be sure to support corner cases (irregular forms, other languages).

nemesisdesign commented on openwisp_controller/config/settings.py:
```diff
@@ -52,3 +52,4 @@ def get_settings_value(option, default):
 }
 HARDWARE_ID_OPTIONS.update(get_settings_value('HARDWARE_ID_OPTIONS', {}))
 HARDWARE_ID_AS_NAME = get_settings_value('HARDWARE_ID_AS_NAME', True)
+DEVICE_VERBOSE_NAME = get_settings_value('DEVICE_VERBOSE_NAME', 'Device')
```
```python
DEVICE_VERBOSE_NAME = get_settings_value('DEVICE_VERBOSE_NAME', (_('Device'), _('Devices')))
```

nemesisdesign reviewed (changes requested):
Good progress @codesankalp!
 
Regarding the test, you can try with [override_settings](https://docs.djangoproject.com/en/3.1/topics/testing/tools/#django.test.override_settings).

See my comments below to improve this patch.

codesankalp commented:
https://github.com/codesankalp/openwisp-controller/commit/8b5fbd1465619916ef89e9784d50f8bc48e4020a
Please check this.
The test works for this but I want to know is this method appropriate according to openwisp rules?

NoumbissiValere commented:
@codesankalp I have commented on your commit

codesankalp commented:
@NoumbissiValere I followed the method in your comment but the test fails. :disappointed:

nemesisdesign reviewed (approved):
Thanks for your effort @codesankalp! Testing is not viable. I will finish this.
